### DISCUSSION TITLE
[犀牛鸟-A1] 结项报告:MoE × End-to-End——现有 one-to-one 能否真正免 NMS

### DISCUSSION THREAD

**Discussion by delei-kong** [Tencent/YOLO-Master#297] [Show and tell]
> 课题:A1|MoE × End-to-End
> 阶段:S1(2026-08-25~09-01)+ S2(2026-09-02~09-08)|作者:delei-kong
> 状态:结项报告

---

## 1. 摘要

A1 课题交付了 **one-to-one 免 NMS 路径在检测小模型上的"精度代价 × 延迟收益"实验设计与分析**,以及一套**可训练/可评测/可部署的实验闭环**。

| 指标 | 4k(S1) | 20k(S2) | 判定 |
|---|---|---|---|
| 免 NMS 精度代价(A→C) | -37.9% | **-3.0%** | 代价大头 = 小数据伪影,固有代价约 3~5% |
| MoE 精度影响(A→B) | -18.0% | **-1.3%** | 4k 负收益是小数据伪影;20k 打平 |
| 免 NMS 下 MoE(C→D) | -12.2% | -5.2% | 收窄过半,交互代价仍存 |
| 免 NMS 延迟收益(GPU) | — | 单图 -5.2% / 批处理 **-23.9%** | 批处理口径交易转正;波动 CV 减半 |

**目标达成:P0 ✅、P1 ✅、P2 ⚠️ 部分完成**(详见 §2)。核心叙事:**免 NMS 交易在 GPU 批服务场景成立**(以 3% 精度代价换 24% 吞吐收益);**MoE 在 20k 上参数 ×2.05 暂时未有明显收益、延迟 +193%**,原因分析见第六章,针对该部分内容的下一步规划见第八章。

## 2. 目标达成情况

| 目标 | 状态 | 完成情况 | 完成依据 |
|---|---|---|---|
| **P0 保底**:固定代码版本复现 detect 基线与 end-to-end 路径;记录 mAP50-95、GPU/CPU batch=1 延迟及标准差 | ✅ 完成 | 四格(2×2)训练/推理闭环,任务书锁 SHA 可复现;指标全口径记录(4k/20k 双规模) | §4.1、§4.2;commit `3fabf78`~`4a48894` |
| **P1 预期**:补齐可训练、可评测、可导出的免 NMS 闭环;2×2 消融(mAP 掉点 ≤2 + CPU 延迟可测改善) | ✅ 完成 | 闭环三件套 SOP 固定;2×2 消融表双规模齐;判据数字:1k 口径 D vs A 绝对差 **0.0200(恰好 2.0pp 达标)**、5k 口径 0.0237(2.4pp 略超,110k 终审);延迟改善 GPU 口径成立、CPU 趋势一致 | §4.1~§4.3、SOP;commit `cb56208`~`c3612a3` |
| **P2 理想**:扩展到 seg/pose,或有证据的负结果(梯度稀疏、匹配冲突、路由坍塌等机制分析) | ⚠️ 部分 | **已完成**:有证据的负结果 + 机制分析(延迟三层归因、路由三步分析、MoE 参数红利未兑现归因);**未完成**:seg/pose 扩展、梯度稀疏与匹配冲突诊断 | §4.4、§4.5、§6;commit `9677f84`、`c3612a3` |

## 3. 实验设计与口径

**2×2 消融矩阵实验**:

| | MoE off(yolo26) | MoE on(yolo26-master-n) |
|---|---|---|
| **NMS on**(end2end=False) | A(基线) | B |
| **NMS off**(免 NMS,one2one) | C | D |

- **训练参数**:epochs 100、imgsz 640、batch 8、seed 0、从头训练;四格唯一差异 = model 参数。
    - S1 数据 = val2017 拆分 4k train / 1k val;
    - S2 数据 = train2017 抽样 20k train / 1k val(同拆分,跨规模可比)+ 训后补跑 5k 全量(官方口径)。唯一变量 = 数据规模。
- **延迟评测**:RTX 4090、FP32、标准部署形态(fuse + torchvision NMS 主口径); 单图(batch=1)与批处理(batch=8)双口径; 固定评测清单(四格共享同一批图);5k 逐批原始日志(seed=0 打乱,四格同 batch 序列,conf=0.001/0.25 双口径)。
- **路由/归因分析**:推理态(专家使用分布、行为差异)+ 训练态(断点序列演化);profiler + 消融归因(见 §4.4/§4.5)。
- **复现**:代码 SHA(训练 `1804866`,评测/分析脚本 `95962c6`);数据与清单存档共享盘;单 seed 局限声明见 §7。

### 4.1 2×2 消融精度

| 格 | 配置 | 1k val best(S1 4k) | 1k val best(S2 20k) | 5k 全量(20k) |
|---|---|---|---|---|
| A | MoE off + NMS | 0.0317 | **0.2506** | 0.2495 |
| B | MoE on + NMS | 0.0260 | 0.2473 | 0.2441 |
| C | MoE off + 免 NMS | 0.0197 | 0.2432 | 0.2371 |
| D | MoE on + 免 NMS | 0.0173 | 0.2306 | 0.2258 |

(5k 全量 P/R:A 0.518/0.362、B 0.500/0.358、C 0.494/0.341、D 0.471/0.337)

四格实验训练过程中的 metrics 曲线：
<img width="1941" height="734" alt="image" src="https://github.com/user-attachments/assets/5b9a22d7-a80d-42f0-8a22-12744160165e" />
四格实验训练中的 loss 曲线：
<img width="2068" height="735" alt="image-1" src="https://github.com/user-attachments/assets/d90c53fb-7151-41ef-9f2d-5a3a6994e3ef" />

(图源:wandb 项目 [`delei-kong-szu/-root-workspace-docs-experiments-2026-09-02-runs`](https://wandb.ai/delei-kong-szu/-root-workspace-docs-experiments-2026-09-02-runs);)

### 4.2 延迟与波动(20k 权重)

由于 NMS-free 设计是为了减少 NMS 算法在不同复杂度下的延迟波动，所以我们设计了相关的 latency 与波动实验：

| 格 | 单图 ms(GPU) | 批处理均摊/图 ms | CV | 5k 逐批均摊(conf=0.25) |
|---|---|---|---|---|
| A | 7.62±0.91 | 1.21 | 0.139 | 5.14 |
| B | 22.35±0.29 | 6.90 | 0.035 | 10.85 |
| C | 7.22±0.26 | **0.92** | **0.036** | **4.91** |
| D | 22.60±0.62 | 6.72 | 0.035 | 10.65 |

CPU(median,共享容器争抢、趋势可读):A 207 / B 1292 / C 206 / D 1398 ms。

<img width="1350" height="900" alt="latency_distribution" src="https://github.com/user-attachments/assets/9b3d4b73-97db-4b18-97ef-2eea62d7d75c" />

<img width="1350" height="1350" alt="latency_vs_gt_n" src="https://github.com/user-attachments/assets/fd6228bd-a348-4f73-b64f-8c2ffbcde587" />

### 4.3 部署导出 parity(20k 四格)

FP32 ONNX(imgsz 640,opset 13),PT vs ORT 同输入输出对比:

| 格 | 最大绝对误差 | 行级匹配(<1e-2) | 判定 |
|---|---|---|---|
| A | 0.0006 | 100.0% | ✅ |
| B | 0.0006 | 100.0% | ✅ |
| C | 0.0001 | 99.3%* | ✅ |
| D | 0.0001 | 100.0% | ✅ |

### 4.4 MoE 延迟归因(B 格 20.6ms 去向)

由于 MOE 模块的 latency 较长，不符合我们的预期，所以深入分析 MOE 各个模块的耗时情况

| 账目 | 数字 | 结论 |
|---|---|---|
| wall | 20.6 ms | 比 dense(A)慢 3 倍 |
| GPU 纯计算 | 6.9 ms(33%) | bmm+softmax(attention)3.8ms、conv 1.2ms |
| CPU 驱动 | 26.9 ms CPU 时间 | **主瓶颈**——kernel 发射与逐专家 Python 调度,GPU 在等 |
| isfinite 同步 | 0.7~1.0 ms | 消融实测 |

发现实际上由于 MOE 中存在 token 的分发机制，需要 CPU 与 GPU 之间进行通信和交互，导致 GPU 同步耗时较长。

### 4.5 路由机制分析(三步)

| 层 | ① 使用集中度(ep100) | ② 专家间相似度 | ③ 集中度演化(ep10→100) |
|---|---|---|---|
| P3(4 专家) | 1.44~1.54(较均衡) | 0.66~0.82(部分分工) | 1.26~1.78 → 1.44~1.52 |
| P4(8 专家) | 1.92~2.28(偏向) | 0.35~0.46(分工明显) | 3.04 → 1.92~1.96 |
| P5(16 专家) | 2.24~3.20(失衡) | **0.14~0.26(分工最强)** | **4.08 → 2.24~3.20** |

<img width="1350" height="900" alt="routing_evolution" src="https://github.com/user-attachments/assets/0997543e-02a8-4ef7-bbaa-92066c44715a" />

专家分工越深(② 越小的层),使用集中度演化越慢(③ 末期仍下行)——分工已形成、路由尚未收敛。说明 MOE 的效果不好可能是由于训练不到位，MOE 尚未收敛导致的。

## 5. 结论:免 NMS 交易

1. **代价大头是小数据伪影**:免 NMS 的 mAP 代价从 4k 的 -37.9% 收敛到 20k 的 -3.0%——4k 上"漏检主导"的代价大部分是数据不足学不动 one2one 匹配所致,固有代价约 3~5%。
2. **交易在 GPU 批服务场景成立**:批处理口径免 NMS 收益 -23.9% 远超 3% 精度代价;单图口径 -5.2% 与代价打平略赚。CPU 口径趋势与 GPU 一致。
3. **免 NMS 削延迟长尾**:波动 CV 0.139→0.066(免 NMS 模型延迟与候选数解耦,corr -0.01 vs NMS 的 +0.80)。
4. **判据边界**:D vs A 掉点 1k 口径恰好 2.0pp 达标、5k 口径 2.4pp 略超——110k 终审锁定。

## 6. 结论:MoE 负结果与归因

1. **现象**:参数 ×2.05(2.42M→4.96M)、计算 ×1.35(5.49→7.42 GFLOPs),精度打平(-1.3%),延迟 +193%。
2. **三层归因**:
   - 主因(假设):**容量冗余**——2.4M 小模型 + 20k 数据未到"容量即瓶颈",2.5M 新增参数无处兑现(裁决实验:同参数量 dense 对照,未跑)
   - 机制层:**路由未收敛**——专家分工已形成(深层最强),但路由使用集中度随 epoch 单调下降、100ep 末期仍在下行,"闲置专家"是中间态而非终态;balance loss 值饱和但分布仍在演化
   - 实现层:**稀疏调度成本**——逐专家 Python 循环 + gather/scatter 导致 CPU 驱动成为延迟主瓶颈
3. **乐观推论(H1)**:更长训练或更高路由 lr 下,P5 层 43% 的闲置参数可能被激活兑现——已启动 S3 检验(§8)。

## 7. 局限声明

- 全部实验单 seed(seed=0);消融差值大于单 seed 可预期噪声,但统计显著性未经过多 seed 检验
- 多 seed 决策:降级为"边界结论的定向保险"——若 110k 终验后判据数字仍在 2pp 边界,仅对 A/D 两格补 2-3 seed
- 4k 模型在 val2017 上训练(泄漏口径),跨规模对比使用 1k 口径
- CPU 延迟绝对值受共享容器进程争抢影响,趋势可读

## 8. 下一步计划

**后续计划(S3,进行中;规划文档:`docs/discuss/A1-S3-MoE训练时长检验与110k决策.md`)**

实验一:**B/D 续训 100→200ep,检验 H1 vs H2**
- 动机(由 §4.5 路由三步分析产生):MoE 参数红利未兑现,是训练时长不足(H1:"MoE 有效只是训练更慢"),还是收敛后仍无优势(H2)?
- 执行:B、D 两格从断点 resume 续训至 200ep
- 观察 ①:路由使用集中度 0~200ep 完整演化曲线——找集中度收敛点(100ep 时仍在下降)
- 观察 ②:每 epoch 1k val mAP 与同范式 dense 的差距走势(B vs A 0.2506、D vs C 0.2432)——收窄还是扩大
- 当前进度:B 格已续训至 ep123(1k val 0.2512,已超 A 基线 0.2506);

实验二:分支决策
- **分支 A(H1 成立)→ 110k 最终训练**:四格全跑、四张任务书(48G 机器 batch 可上调,记录口径变化;epochs 依实验一的收敛观察届时再定)
- **分支 B(H2 成立)→ 负结果收尾**:不扩大数据;MoE 线按负结果呈现(实现层候选解释:批量专家计算变体),主线转向免 NMS 叙事收口

其他候选(暂缓):
- 同参数量(~5M)dense 对照——"容量冗余"主因的裁决实验,视 H1/H2 结论再议
- MoE 批量专家计算(BatchedExpertComputation)实现升级 + 延迟账本重测
- 规模收敛图(4k→20k→110k 三连点)待 110k 数字后绘制

## 9. 资产清单

- 代码:12 个 commit(PR `pr/a1-s2`,rebase 于 main;目录规范:`scripts/a1/`、`ultralytics/cfg/models/26/`、`ultralytics/cfg/datasets/`)
- 脚本:评测 pipeline / 逐批日志 / 延迟归因 / 路由三步 / 导出 parity / 数据准备,共 16 个
- 报告与原始数据: [`阶段报告`](https://github.com/Tencent/YOLO-Master/discussions/254)(阶段报告、SOP、延迟账本、归因、路由分析、逐批日志、图表)
